CRM/Activity/Form/Task.php

```diff
@@ -116,10 +116,16 @@ public static function preProcessCommon(&$form, $useTable = FALSE) {
 
       // CRM-12675
       $activityClause = NULL;
-      if (! CRM_Core_Permission::check('access CiviContribute')) {
-        $activityClause = ' (activity_type.component_id IS NULL OR activity_type.component_id <> 2) ';
-      }
 
+      $components = CRM_Core_Component::getNames();
+      foreach ($components as $componentID => $componentName) {
+        if (!CRM_Core_Permission::check("access $componentName")) {
+          $componentClause = " (activity_type.component_id IS NULL OR activity_type.component_id <> {$componentID}) ";
+        }
+      }
+      if (!empty($componentClause)) {
+        $activityClause = implode(' AND ', $componentClause);
+      }
```

Comment:
This code appears to be overwriting the same variable over and over, which means if the user lacks access to more than one component, it will only block the last (if at all, I'm not sure exactly what `implode` will do with a string).
I think what was meant was to declare $componentClause as an array, and then add items to it, rather than overwriting a string.